### PR TITLE
[workloadmeta] Add kubernetes namespace information to workloadmeta

### PR DESCRIPTION
**Pull request by jennchenn**
### What does this PR do?

Adds a new `KubernetesNamespace` type to workloadmeta to collect and store kubernetes namespace information.

### Motivation

Among other features, this will be used to add `kubernetes_namespace_labels_as_tags` to kubernetes events.

### Additional Notes

Currently namespace information is only collected if the config `kubernetes_namespace_collection_enabled` is set to `true`.

### Describe how to test/QA your changes

1. Deploy the agent and cluster agent
2. Exec into the cluster agent and verify that you do not see any `kubernetes_namespace` entries in workloadmeta (when running `agent workload-list -v`)
3. Update the configuration to set `kubernetes_namespace_collection_enabled` = `true`
4. Redeploy and verify that you see an entry in workloadmeta for each namespace
```
root@datadog-agent-cluster-agent-695595b86f-wzp6h:/# agent workload-list -v
2024-04-30 18:04:22 UTC | CLUSTER | DEBUG | (pkg/util/log/log.go:791 in func1) | 'use_proxy_for_cloud_metadata' is enabled: adding cloud provider URL to the no_proxy list
2024-04-30 18:04:22 UTC | CLUSTER | DEBUG | (pkg/util/log/log.go:786 in func1) | FIPS mode is disabled
2024-04-30 18:04:22 UTC | CLUSTER | INFO | (pkg/util/log/log.go:831 in func1) | 2 Features detected from environment: kubernetes,kube_orchestratorexplorer

=== Entity kubernetes_node sources(merged):[kubeapiserver] id: jenn-display-name ===
----------- Entity ID -----------
Kind: kubernetes_node ID: jenn-display-name

----------- Entity Meta -----------
Name: jenn-display-name
Namespace:
Annotations: kubeadm.alpha.kubernetes.io/cri-socket:unix:///var/run/cri-dockerd.sock node.alpha.kubernetes.io/ttl:0 volumes.kubernetes.io/controller-managed-attach-detach:true
Labels: minikube.k8s.io/primary:true minikube.k8s.io/updated_at:2024_04_29T17_26_35_0700 minikube.k8s.io/version:v1.32.0 node-role.kubernetes.io/control-plane: beta.kubernetes.io/os:linux minikube.k8s.io/commit:***********************************0512d beta.kubernetes.io/arch:arm64 minikube.k8s.io/name:jenn-display-name kubernetes.io/arch:arm64 kubernetes.io/hostname:jenn-display-name kubernetes.io/os:linux node.kubernetes.io/exclude-from-external-load-balancers:
===

=== Entity kubernetes_namespace sources(merged):[kubeapiserver] id: default ===
----------- Entity ID -----------
Kind: kubernetes_namespace ID: default

----------- Entity Meta -----------
Name: default
Namespace:
Annotations:
Labels: kubernetes.io/metadata.name:default
===

=== Entity kubernetes_namespace sources(merged):[kubeapiserver] id: kube-node-lease ===
----------- Entity ID -----------
Kind: kubernetes_namespace ID: kube-node-lease

----------- Entity Meta -----------
Name: kube-node-lease
Namespace:
Annotations:
Labels: kubernetes.io/metadata.name:kube-node-lease
===

=== Entity kubernetes_namespace sources(merged):[kubeapiserver] id: kube-public ===
----------- Entity ID -----------
Kind: kubernetes_namespace ID: kube-public

----------- Entity Meta -----------
Name: kube-public
Namespace:
Annotations:
Labels: kubernetes.io/metadata.name:kube-public
===

=== Entity kubernetes_namespace sources(merged):[kubeapiserver] id: kube-system ===
----------- Entity ID -----------
Kind: kubernetes_namespace ID: kube-system

----------- Entity Meta -----------
Name: kube-system
Namespace:
Annotations:
Labels: kubernetes.io/metadata.name:kube-system
===

=== Entity kubernetes_namespace sources(merged):[kubeapiserver] id: test ===
----------- Entity ID -----------
Kind: kubernetes_namespace ID: test

----------- Entity Meta -----------
Name: test
Namespace:
Annotations:
Labels: kubernetes.io/metadata.name:test
===
```
5. Add a label to one of the namespaces and verify that this change is reflected in wlm
```
$ k label namespace test app=test

agent workload-list -v
... 
=== Entity kubernetes_namespace sources(merged):[kubeapiserver] id: test ===
----------- Entity ID -----------
Kind: kubernetes_namespace ID: test

----------- Entity Meta -----------
Name: test
Namespace:
Annotations:
Labels: kubernetes.io/metadata.name:test app:test
===
```

### CODE REVIEW

File: comp/core/tagger/taggerimpl/collectors/workloadmeta_extract.go [hunk 1]
```diff
@@ -435,6 +437,28 @@ func (c *WorkloadMetaCollector) handleKubeNode(ev workloadmeta.Event) []*types.T
 	return tagInfos
 }
 
+func (c *WorkloadMetaCollector) handleKubeNamespace(ev workloadmeta.Event) []*types.TagInfo {
+	namespace := ev.Entity.(*workloadmeta.KubernetesNamespace)
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/collectors/workloadmeta_extract.go:441]: Question:

`ev` is only used once to cast the entity to a namespace object.

after that, the `namespace` object is used only once below to build the tagger entity ID (line 451). 

I don't see where the entity data (i.e. the namespace labels and annotations) are being used in the tagger. To me it seems like the workloadmeta extractor in the tagger is always setting empty tag lists for all cardinalities for namespaces.

Am I missing something?

File: comp/core/workloadmeta/store_test.go [hunk 2]
```diff
@@ -679,6 +679,49 @@ func TestGetKubernetesDeployment(t *testing.T) {
...
+		t.Errorf("expected namespace %q to match the one in the store", retrievedNamespace.ID)
+	}
+
+	s.handleEvents([]CollectorEvent{
+		{
+			Type:   EventTypeUnset,
+			Source: fooSource,
+			Entity: namespace,
+		},
+	})
```
Comment (adel121) [inline comp/core/workloadmeta/store_test.go:719]: Question

After this we are asserting that the `datadog-cluster-agent` namespace doesn't exist. 
So why are we handling the same event again here? Is this necessary?

**Review by adel121 — COMMENTED** [review]
Thanks @jennchenn 

Left 2 questions

File: comp/core/tagger/taggerimpl/collectors/workloadmeta_extract.go (hunk 1 above)
Comment (jennchenn) [reply comp/core/tagger/taggerimpl/collectors/workloadmeta_extract.go]: Removed the changes in this file for now, to be added in again later (with actual functionality)

File: comp/core/workloadmeta/store_test.go (hunk 2 above)
Comment (jennchenn) [reply comp/core/workloadmeta/store_test.go]: good point, removed!

File: comp/core/tagger/taggerimpl/collectors/workloadmeta_extract.go (hunk 1 above)
Comment (davidor) [reply comp/core/tagger/taggerimpl/collectors/workloadmeta_extract.go]: If you remove `case workloadmeta.KindKubernetesNamespace:` the `default` will apply which logs an error.
I think it's better to do what we're doing with deployments and processes:
```go
case workloadmeta.KindKubernetesDeployment:
// tagInfos = append(tagInfos, c.handleDeployment(ev)...) No tags for now
```
That way the agent will not log unnecessary errors.

**Review by Kaderinho — APPROVED** [review]
LGTM for the config. QQ: is this customer facing ? If yes, why don't we add a release note ?

**Comment by jennchenn** [comment]
> LGTM for the config. QQ: is this customer facing ? If yes, why don't we add a release note ?

thanks for reviewing! most of these changes are internal (for now) but i did add this config which will be exposed to the customer - if they were to enable it they shouldn't notice much change at the moment but will in the future as i continue to build out this feature. however since there is a new config being exposed to customers, would it make sense to add a changelog entry for it? 

alternatively i can remove the config from this PR and only add it in with other customer-facing changes

**Review by davidor — APPROVED** [review]
Looks good to me.
We just need to keep in mind that some things, like the config param and maybe some structs, may change when @adel121 implements the generic collector.

**Review by adel121 — APPROVED** [review]
LGTM

thanks @jennchenn

**Pull request merged**